plutus-ledger/src/Ledger/Typed/Scripts.hs

```diff
@@ -1,9 +1,116 @@
+{-# LANGUAGE DerivingStrategies   #-}
+{-# LANGUAGE GADTs                #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE UndecidableInstances #-}
+
 {-# OPTIONS_GHC -Wno-missing-import-lists #-}
 
 module Ledger.Typed.Scripts
-    ( module Export
-    ) where
+  ( module Export
+  , TypedScriptTxIn (tyTxInTxIn, tyTxInOutRef)
+  , makeTypedScriptTxIn
+  , txInValue
+  , makePubKeyTxIn
+  , typePubKeyTxIn
+  , typeScriptTxIn
+  ) where
 
+import Control.Monad.Except (MonadError (throwError))
+import Data.Aeson (FromJSON (..), ToJSON (..))
+import GHC.Generics (Generic)
+import Ledger.Tx.Internal (LedgerPlutusVersion, TxIn (TxIn, txInRef, txInType),
+                           TxInType (ConsumePublicKeyAddress, ConsumeScriptAddress))
 import Ledger.Typed.Scripts.Orphans as Export ()
 import Plutus.Script.Utils.V1.Typed.Scripts as Export
 import Plutus.Script.Utils.V1.Typed.TypeUtils as Export
+import Plutus.V1.Ledger.Api (Datum (Datum), FromData, Redeemer (Redeemer), ToData (..), TxOut (txOutValue), TxOutRef,
+                             Value)
+
+-- | A 'TxIn' tagged by two phantom types: a list of the types of the data scripts in the transaction; and the connection type of the input.
+data TypedScriptTxIn a = TypedScriptTxIn
+  { tyTxInTxIn   :: TxIn,
+    tyTxInOutRef :: TypedScriptTxOutRef a
+  }
+
+instance Eq (DatumType a) => Eq (TypedScriptTxIn a) where
+  l == r =
+    tyTxInTxIn l == tyTxInTxIn r
+      && tyTxInOutRef l == tyTxInOutRef r
+
+-- instance (FromJSON (DatumType a), FromData (DatumType a), ToData (DatumType a)) => FromJSON (TypedScriptTxIn a) where
+--   parseJSON (Data.Aeson.Object v) =
+--     TypedScriptTxIn <$> v .: "tyTxInTxIn" <*> v .: "tyTxInOutRef"
+--   parseJSON invalid = typeMismatch "Object" invalid
+
+-- instance (ToJSON (DatumType a)) => ToJSON (TypedScriptTxIn a) where
+--   toJSON TypedScriptTxIn {tyTxInTxIn, tyTxInOutRef} =
+--     object ["tyTxInTxIn" .= tyTxInTxIn, "tyTxInOutRef" .= tyTxInOutRef]
```

Comment:
Can be removed?

Reply:
No this still needs to be fixed, but that involves some more changes in this PR is already too big. Will make a ticket for it.

Reply:
Seems like it was removed in a recent commit. I don't see this commented code in the PR. Shouldn't we keep it?

Reply:
No, after some more thought it seems we don't need it.